Numerical_solution_of_SEIRD.Rmd

```diff
@@ -0,0 +1,356 @@
+  t=numerical_solution_dense[,1],
+  f=numerical_solution_dense[,2],
+  method="Numerical, Δt=0.1"
+)
+
+results_true <- data.frame(
+  t=seq(0, 3, by=0.01),
+  f=true_solution(seq(0, 3, by=0.01)),
+  method="True solution"
+)
+
+
+ggplot() +
+  geom_point(data=rbind(results_sparse, results_dense),
+             aes(x=t, y=f, color=method),
+             show.legend=F) +
+  geom_line(data=rbind(results_sparse, results_dense, results_true),
+            aes(x=t, y=f, color=method))
+```
+
+We observe that the large step size of $\Delta t=1$ causes significant
+inaccuracy in the solution. In particular, because the derivative is increasing
+over time, the forward Euler method (which, between solver time steps, assumes
+that the  derivative remains constant at an old value) produces an approximate
+solution that is too low. For the smaller time step $\Delta t=0.1$, the
+approximate solution is more accurate, although a negative deviation is still
+apparent.
+
+## Effect of solver step size on the SEIR model
+
+First, we load the SEIRD model from comomodels, and specify plausible parameter
+values and initial conditions.
+
+```{r}
+model <- comomodels::SEIRD()
+transmission_parameters(model) <- list(beta=1.0, kappa=0.9, gamma=0.5, mu=0.1)
+initial_conditions(model) <- list(S0=0.999, E0=0, I0=0.001, R0=0)
+```
+
+Next, we run the model using the forward Euler method. We investigate two 
+different choices for the solver step size. The first is a daily time step,
+which naively might be considered an acceptable default choice, given that we 
+are interested in simulating daily cases and deaths. We also consider a much 
+smaller time step of 0.001 days, for which we know that the solution is 
+sufficiently accurate (see the following section for further details on 
+obtaining accurate numerical solutions).
+
+```{r, fig.width=7, fig.height=5}
+times_daily <- seq(0, 50, by=1)
+solution_daily <- run(model, times_daily, solve_method="euler")
+changes_daily <- solution_daily$changes
+
+times_dense <- seq(0, 50, by=0.001)
+solution_dense <- run(model, times_dense, solve_method="euler")
+df <- solution_dense$changes
+df <- as.data.frame(df)
+
+# Sum the deaths and incidences from the dense output to obtain daily values
+changes_dense <- data.frame(time=NULL, value=NULL, compartment=NULL)
+for (t in times_daily) {
+  cases <- subset(df, t-1<df[,1] & df[,1]<=t)
+  incidence <- subset(cases, cases[,3]=="Incidence")
+  death <- subset(cases, cases[,3]=="Deaths")
+  changes_dense <- rbind(changes_dense, 
+                         list(time=t, 
+                              value=sum(incidence$value), 
+                              compartment="Incidence", 
+                              age_range="0-150"))
+  changes_dense <- rbind(changes_dense, 
+                         list(time=t, 
+                              value=sum(death$value), 
+                              compartment="Deaths", 
+                              age_range="0-150"))
+}
```

Comment:
In this bit of code, it gives lots of warning, can we suppress that?

![image](https://user-images.githubusercontent.com/26725646/126598678-fc0034d0-e2b9-4139-a1b2-4c19f9fac512.png)

Reply:
Interestingly I don't get these warnings. maybe a version difference? I've made a change that could fix this (rbind a dataframe instead of list), so let me know if you still see the warnings.

Reply:
You are most probably right that it's just version difference. Yeah, thanks for fixing it, the warnings are gone now.